File: examples/csharp/PortPicker.cs
using System.IO.Ports;

namespace MavlinkSitlGcs;

/// <summary>Lists available serial ports and reads a selection from stdin.</summary>
public static class PortPicker
{
    public static string PickSerialPort()
    {
        var portNames = SerialPort.GetPortNames();
        if (portNames.Length == 0)
        {
            throw new InvalidOperationException(
                "No serial ports found. Connect SITL or a USB adapter.");
        }

        Array.Sort(portNames);

        Console.WriteLine();
        Console.WriteLine("Available serial ports:");
        for (var index = 0; index < portNames.Length; index++)
        {
            Console.WriteLine($"  [{index}] {portNames[index]}");
        }

        Console.WriteLine();
        Console.Write($"Select port [0-{portNames.Length - 1}]: ");

        var line = Console.ReadLine()?.Trim();
        if (string.IsNullOrEmpty(line))
        {
            throw new InvalidOperationException("Port selection required");
        }

        if (!int.TryParse(line, out var selected) || selected < 0 || selected >= portNames.Length)
        {
            throw new InvalidOperationException($"Invalid port selection: {line}");
        }

        var portName = portNames[selected];
        Console.WriteLine($"Selected {portName}");
        return portName;
    }

    /// <summary>Parse <c>--baud &lt;rate&gt;</c> from CLI arguments (default 57600).</summary>
    public static int ParseBaudRate(string[] args, int defaultBaud = 57600)
    {
        for (var index = 0; index < args.Length - 1; index++)
        {
            if (args[index] == "--baud")
            {
                if (!int.TryParse(args[index + 1], out var value) || value <= 0)
                {
                    throw new ArgumentException($"Invalid --baud value: {args[index + 1]}");
                }

                return value;
            }
        }

        return defaultBaud;
    }
}

File: examples/rust/src/port_picker.rs
//! Serial port discovery and CLI baud-rate parsing.

use std::io::{self, Write};

/// Lists available serial ports and reads a selection from stdin.
pub fn pick_serial_port() -> io::Result<String> {
    let ports = serialport::available_ports().map_err(|error| {
        io::Error::new(
            io::ErrorKind::NotFound,
            format!("failed to enumerate serial ports: {error}"),
        )
    })?;

    if ports.is_empty() {
        return Err(io::Error::new(
            io::ErrorKind::NotFound,
            "No serial ports found. Connect SITL or a USB adapter.",
        ));
    }

    println!();
    println!("Available serial ports:");
    for (index, port) in ports.iter().enumerate() {
        let details = match &port.port_type {
            serialport::SerialPortType::UsbPort(info) => {
                let mut parts = Vec::new();
                if let Some(manufacturer) = &info.manufacturer {
                    parts.push(manufacturer.clone());
                }
                if let Some(product) = &info.product {
                    parts.push(product.clone());
                }
                parts.join(" — ")
            }
            serialport::SerialPortType::BluetoothPort => "Bluetooth".to_string(),
            serialport::SerialPortType::PciPort => "PCI".to_string(),
            serialport::SerialPortType::Unknown => String::new(),
        };
        if details.is_empty() {
            println!("  [{index}] {}", port.port_name);
        } else {
            println!("  [{index}] {} ({details})", port.port_name);
        }
    }
    println!();
    print!("Select port [0-{}]: ", ports.len() - 1);
    io::stdout().flush()?;

    let mut line = String::new();
    io::stdin().read_line(&mut line)?;
    let line = line.trim();
    if line.is_empty() {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "Port selection required",
        ));
    }

    let selected: usize = line.parse().map_err(|_| {
        io::Error::new(
            io::ErrorKind::InvalidInput,
            format!("Invalid port selection: {line}"),
        )
    })?;
    if selected >= ports.len() {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            format!("Invalid port selection: {line}"),
        ));
    }

    let port_name = ports[selected].port_name.clone();
    println!("Selected {port_name}");
    Ok(port_name)
}

/// Parse `--baud <rate>` from CLI arguments (default 57600).
pub fn parse_baud_rate(args: &[String], default_baud: u32) -> Result<u32, String> {
    for index in 0..args.len().saturating_sub(1) {
        if args[index] == "--baud" {
            let value = args[index + 1]
                .parse::<u32>()
                .map_err(|_| format!("Invalid --baud value: {}", args[index + 1]))?;
            if value == 0 {
                return Err(format!("Invalid --baud value: {}", args[index + 1]));
            }
            return Ok(value);
        }
    }
    Ok(default_baud)
}
